### PR TITLE
fix: updates from review

### PR DESCRIPTION
**Pull request by Saransh-cpp**
XRef #991 

- a new util function to register projection classes
- do not modify the global awkward behavior
- tests for geometric dimension errors

cc: @nsmith- @lgray

### CODE REVIEW

File: src/coffea/nanoevents/methods/candidate.py [hunk 1]
```diff
@@ -70,23 +71,29 @@ class PtEtaPhiECandidate(Candidate, vector.PtEtaPhiELorentzVector):
...
+    CandidateArray,  # noqa: F821
+)
+register_projection_classes(
+    PtEtaPhiMCandidateArray,  # noqa: F821
+    vector.PolarTwoVectorArray,
+    vector.SphericalThreeVectorArray,
+    vector.LorentzVectorArray,
+    PtEtaPhiMCandidateArray,  # noqa: F821
+)
+register_projection_classes(
```
Comment (nsmith-) [inline src/coffea/nanoevents/methods/candidate.py:90]: Ok this is my misunderstanding, I hadn't realized each case is somewhat different so the function doesn't in the end save any code duplication. Though at least for many of these the first 3 arguments are the same

Comment (nsmith-) [reply src/coffea/nanoevents/methods/candidate.py:90]: I think in this case what you had before was more readable code, is it ok to revert? No need to make separate PR

Comment (lgray) [reply src/coffea/nanoevents/methods/candidate.py:90]: arg, sorry, I thought convo was done, my bad

Comment (Saransh-cpp) [reply src/coffea/nanoevents/methods/candidate.py:90]: I'll have to create a separate "revert" PR to fix this. There are some changes in this PR that should go in, like the extra tests and not modifying the global awkward behavior.

Comment (nsmith-) [reply src/coffea/nanoevents/methods/candidate.py:90]: Feel free to revert the subset of changes directly on the other PR and push, or make a revert PR if you prefer

**Pull request merged**